ios/keyman/Keyman/Keyman/MainViewController.swift

```diff
@@ -0,0 +1,1200 @@
+    activityView.center = CGPoint(x: containerView.bounds.size.width * 0.5, y: containerView.bounds.size.height * 0.5)
+    activityView.startAnimating()
+    containerView.addSubview(activityView)
+    overlayWindow.addSubview(containerView)
+  }
+
+  func dismissActivityIndicator() {
+    overlayWindow.viewWithTag(activityIndicatorViewTag)?.removeFromSuperview()
+    overlayWindow.isHidden = true
+  }
+
+  func sliderValueChanged(_ sender: UISlider) {
+    textSize = CGFloat(sender.value + Float(minTextSize)).rounded()
+    textView.font = textView.font?.withSize(textSize)
+    let textSizeTitle = presentedViewController?.view.viewWithTag(2) as? UILabel
+    textSizeTitle?.text = "Text size: \(Int(textSize))"
+  }
+
+  func rectForBarButtonItem(_ barButtonItem: UIBarButtonItem) -> CGRect {
+    let view =  barButtonItem.value(forKey: "view") as? UIView
+    return view?.frame ?? CGRect.zero
+  }
+
+  func startTimer() {
+    if loadTimer == nil {
+      loadTimer = Timer.scheduledTimer(timeInterval: 1.0, target: self,
+                                       selector: #selector(self.timerAction), userInfo: nil, repeats: true)
+    }
+  }
+
+  func stopTimer() {
+    if let timer = loadTimer {
+      timer.invalidate()
+      loadTimer = nil
+    }
+  }
+
+  func timerAction() {
+    if updateStatus == 1 {
+      stopTimer()
+
+      dismissActivityIndicator()
+
+      textView.isEditable = true
+      textView.becomeFirstResponder()
+
+      if let launchUrl = launchUrl {
+        performAction(from: launchUrl)
+      } else {
+        loadSavedUserText()
+        if shouldShowGetStarted() {
+          perform(#selector(self.showGetStartedView), with: nil, afterDelay: 1.0)
+        }
+      }
+    } else if updateStatus == -1 {
+      stopTimer()
+      dismissActivityIndicator()
+      loadSavedUserText()
+      textView.isEditable = true
+      textView.becomeFirstResponder()
+      if shouldShowGetStarted() {
+        perform(#selector(self.showGetStartedView), with: nil, afterDelay: 1.0)
+      }
+    }
+  }
+
+  func resetTextViewCursor() {
+    textView.selectedRange = NSRange(location: 0, length: 0)
+
+    // TODO: Figure out what this does
+    if let size = textView.font?.pointSize {
+      textView.font = textView.font?.withSize(size + 5)
+      textView.font = textView.font?.withSize(size - 5)
+    }
+  }
+
+  func loadSavedUserText() {
+    let userData = AppDelegate.activeUserDefaults()
+    let userText = userData.object(forKey: userTextKey) as? String
+    let userTextSize = userData.object(forKey: userTextSizeKey) as? String
+
+    if let text = userText, !text.isEmpty {
+      textView.text = text
+    }
+
+    if let sizeString = userTextSize, let size = Float(sizeString) {
+      textSize = CGFloat(size)
+      textView.font = textView.font?.withSize(textSize)
+      textSizeController.value = (Float(textSize - minTextSize))
+    }
+  }
+
+  func params(of query: String) -> [String: String] {
```

Comment:
A new function?  Seems to be a helper for the old function that follows it.

Reply:
Yeah I couldn't help but try to get rid of the looping and substringing that was going on to get the URL params.